project/apps/user-service/src/adapters/db/users.supabase.ts

```diff
@@ -21,5 +28,166 @@ export class SupabaseUsersRepository implements UsersRepository {
     this.supabase = createClient(supabaseUrl, supabaseKey);
   }
 
-  // Refer to questions.supabase.ts for examples
+  async findAll(filters: UserFiltersDto): Promise<UserCollectionDto> {
+    const { email, username, offset, limit } = filters;
+
+    let queryBuilder = this.supabase
+      .from(this.PROFILES_TABLE)
+      .select('*', { count: 'exact' });
+
+    if (email || username) {
+      const orFilter = [];
+      if (email) {
+        orFilter.push(`email.eq.${email}`);
+      }
+      if (username) {
+        orFilter.push(`username.eq.${username}`);
+      }
+      if (orFilter.length > 0) {
+        queryBuilder = queryBuilder.or(orFilter.join(','));
+      }
+    }
+
+    const totalCountQuery = queryBuilder;
+
+    let dataQuery = queryBuilder;
+    if (limit) {
+      if (offset) {
+        dataQuery = dataQuery.range(offset, offset + limit - 1); // Supabase range is inclusive
+      } else {
+        dataQuery = dataQuery.limit(limit);
+      }
+    }
+
+    // Execute the data query
+    const { data: users, error } = await dataQuery;
+
+    // Execute the total count query
+    const { count: totalCount, error: totalCountError } = await totalCountQuery;
+
+    if (error || totalCountError) {
+      throw error || totalCountError;
+    }
+
+    const metadata: collectionMetadataDto = {
+      count: users ? users.length : 0,
+      totalCount: totalCount ?? 0,
+    };
+
+    return {
+      metadata,
+      users,
+    } as UserCollectionDto;
+  }
+
+  async findById(id: string): Promise<UserDataDto> {
+    const { data, error } = await this.supabase
+      .from(this.PROFILES_TABLE)
+      .select()
+      .eq('id', id)
+      .single<UserDataDto>();
+
+    if (error) {
+      throw error;
+    }
+
+    return data;
+  }
+
+  async updateById(userDetails: UpdateUserDto): Promise<UserDataDto> {
+    const {
+      id,
+      email: newEmail,
+      username: newUsername,
+      password: newPassword,
+    } = userDetails;
+
+    // Update user email and/or password in auth table
+    const { error: authError } = await this.supabase.auth.updateUser({
+      email: newEmail,
+      password: newPassword,
+    });
+
+    if (authError) {
+      throw authError;
+    }
+
+    // Update user details in profiles table
+    const updateData: any = {};
+    if (newEmail) {
+      updateData.email = newEmail;
+    }
+    if (newUsername) {
+      updateData.username = newUsername;
+    }
+
+    const { data, error } = await this.supabase
+      .from(this.PROFILES_TABLE)
+      .update({
+        email: newEmail,
+        username: newUsername,
+      })
+      .eq('id', id)
+      .select()
+      .single<UserDataDto>();
+
+    if (error) {
+      throw error;
+    }
+    return data;
+  }
+
+  async updatePrivilegeById(id: string): Promise<any> {
+    const user = await this.findById(id);
+    const newRole = user.role == 'Admin' ? 'User' : 'Admin';
```

Comment:
correct me if im wrong but the current implementation for `updatePrivilege` is just a toggle switch right?

in the future should we make it such that the new role needs to be defined in the request instead?

Reply:
yep current implementation is toggle switch. i think in future if there are more roles then it'll be better to define in the request and let client side control, but currently with only admin and user, it should be okay?

Reply:
ok yup good with me 👍